File: sdks/typescript/tests/failover.test.ts
// Failover integration test.
//
// Runs only when `SHARDD_FAILOVER_GATEWAYS` is set — a comma-separated
// list of local gateway URLs. `./run sdk:test:failover` brings up the
// 3-gateway harness, sets the env, and invokes vitest.
import { describe, expect, it } from "vitest";
import { Client } from "../src/index.js";

const raw = process.env.SHARDD_FAILOVER_GATEWAYS;
const bucket = process.env.SHARDD_FAILOVER_BUCKET ?? "failover-test";
const GATEWAYS = raw
  ? raw.split(",").map((s) => s.trim()).filter(Boolean)
  : [];
const ENABLED = GATEWAYS.length > 0;

// vitest's `describe.skip` keeps the suite visible in the report but
// skips each test when the env var is absent — keeps CI output honest.
const maybe = ENABLED ? describe : describe.skip;

maybe("failover against docker harness", () => {
  it("all-healthy: probe picks one, write + idempotent replay succeed", async () => {
    const client = new Client("local-dev", { edges: GATEWAYS });
    const first = await client.createEvent(bucket, "alice", 10, {
      note: "failover test: phase A",
    });
    expect(first.deduplicated).toBe(false);

    const replay = await client.createEvent(bucket, "alice", 10, {
      idempotencyNonce: first.event.idempotency_nonce,
    });
    expect(replay.deduplicated).toBe(true);
    expect(replay.event.event_id).toBe(first.event.event_id);
  });

  it("closed port mixed in: probe skips it", async () => {
    const edges = ["http://127.0.0.1:1", ...GATEWAYS];
    const client = new Client("local-dev", { edges });
    const result = await client.createEvent(bucket, "bob", 5, {
      note: "failover test: phase B",
    });
    expect(result.deduplicated).toBe(false);
  });

  it("single survivor: writes route to the one healthy edge", async () => {
    const [survivor] = GATEWAYS;
    const client = new Client("local-dev", {
      edges: ["http://127.0.0.1:1", "http://127.0.0.1:2", survivor!],
    });
    const result = await client.createEvent(bucket, "carol", 7, {
      note: "failover test: phase C",
    });
    expect(result.deduplicated).toBe(false);
  });

  it(
    "mid-run outage (gated on SHARDD_FAILOVER_KILLED_GATEWAY): writes still succeed",
    async () => {
      if (!process.env.SHARDD_FAILOVER_KILLED_GATEWAY) {
        return; // not in kill-phase of the harness
      }
      const client = new Client("local-dev", { edges: GATEWAYS });
      const result = await client.createEvent(bucket, "dan", 3, {
        note: "failover test: phase D — mid-outage",
      });
      expect(result.deduplicated).toBe(false);
    },
  );
});

File: sdks/kotlin/src/test/kotlin/xyz/shardd/sdk/FailoverTest.kt
package xyz.shardd.sdk

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assumptions.assumeTrue
import org.junit.jupiter.api.Test

class FailoverTest {
    @Test
    fun allHealthyProbePicksOneAndWritesSucceed() {
        val gateways = gateways()
        assumeTrue(gateways.isNotEmpty(), "set SHARDD_FAILOVER_GATEWAYS to run")

        val client = Client("local-dev", ClientOptions(edges = gateways))
        val first =
            client.createEvent(
                bucket = bucket(),
                account = "alice",
                amount = 10,
                opts = CreateEventOptions(note = "failover test: phase A"),
            )
        assertFalse(first.deduplicated)

        val replay =
            client.createEvent(
                bucket = bucket(),
                account = "alice",
                amount = 10,
                opts = CreateEventOptions(idempotencyNonce = first.event.idempotencyNonce),
            )
        assertEquals(first.event.eventId, replay.event.eventId)
        assertTrue(replay.deduplicated)
    }

    @Test
    fun closedPortMixedInIsSkippedByProbe() {
        val gateways = gateways()
        assumeTrue(gateways.isNotEmpty(), "set SHARDD_FAILOVER_GATEWAYS to run")

        val client =
            Client(
                "local-dev",
                ClientOptions(edges = listOf("http://127.0.0.1:1") + gateways),
            )
        val result =
            client.createEvent(
                bucket = bucket(),
                account = "bob",
                amount = 5,
                opts = CreateEventOptions(note = "failover test: phase B"),
            )
        assertFalse(result.deduplicated)
    }

    @Test
    fun singleSurvivorStillSucceeds() {
        val gateways = gateways()
        assumeTrue(gateways.isNotEmpty(), "set SHARDD_FAILOVER_GATEWAYS to run")

        val edges = listOf("http://127.0.0.1:1", "http://127.0.0.1:2", gateways.first())
        val client = Client("local-dev", ClientOptions(edges = edges))
        val result =
            client.createEvent(
                bucket = bucket(),
                account = "carol",
                amount = 7,
                opts = CreateEventOptions(note = "failover test: phase C"),
            )
        assertFalse(result.deduplicated)
    }

    @Test
    fun midTestOutageDoesNotBreakWrites() {
        assumeTrue(System.getenv("SHARDD_FAILOVER_KILLED_GATEWAY") != null, "set SHARDD_FAILOVER_KILLED_GATEWAY to run")
        val gateways = gateways()
        assumeTrue(gateways.isNotEmpty(), "set SHARDD_FAILOVER_GATEWAYS to run")

        val client = Client("local-dev", ClientOptions(edges = gateways))
        val result =
            client.createEvent(
                bucket = bucket(),
                account = "dan",
                amount = 3,
                opts = CreateEventOptions(note = "failover test: phase D - mid-outage"),
            )
        assertFalse(result.deduplicated)
    }

    private fun bucket(): String = System.getenv("SHARDD_FAILOVER_BUCKET") ?: "failover-test"

    private fun gateways(): List<String> =
        System.getenv("SHARDD_FAILOVER_GATEWAYS")
            ?.split(',')
            ?.map { it.trim() }
            ?.filter { it.isNotEmpty() }
            .orEmpty()
}
